### PR TITLE
Feat v3

### PR DESCRIPTION
**Pull request by GalloDaSballo**
Amazing work @rayeaster holy shit 200k !!!! LFGOOOOOOO

Here's some thoughts

## Equivalency Tests
Can we please have tests for equivalency between the non-optimized pricer, and the optimized pricer?

We need to be able to prove that all quotes are the same, while being more gas efficient

Reference Implementation: https://github.com/GalloDaSballo/fair-selling/blob/feat-v3/contracts/archive/FullOnChainPricingMainnet.sol

## Some Comments left in code
Screenshot 2022-07-20 at 23 39 47
Please kill or refactor

## Library Imports vs Copy Paste
I wonder if we can import the UniV3 and Balancer Libraries via Imports.
If they have a Github Release we can just import them that way to avoid the needless copy paste

## Balancer Estimate for Different Curves
Re the Balancer Estimate:
Are we sure this will work with all pools (I think some pools with stable pricing will use a different curve).
Maybe worth:
- Documenting if code is incomplete / only for certain pools
- Figuring out a way to codify multiple pools with different math and supporting them

## Last Level of Optimization
At this time the "heuristic" is to skip a check early
However, those same slots could be used to heuristically compare an approximation of a quote for each price, with the goal of pre-sorting the results (e.g. Bal always beats UniV2 as long as token0 + token1 is greater than X), this way we can re-use the memory stored data we already fetched and gain further information from it.

Before doing this, we must have complete coverage and have proof of equivalence against the V2 Release

### CODE REVIEW

File: contracts/OnChainPricingMainnet.sol [hunk 1]
```diff
@@ -58,7 +62,7 @@ contract OnChainPricingMainnet {
     address public constant UNIV3_QUOTER = 0xb27308f9F90D607463bb33eA1BeBb41C27CE5AB6;
     bytes32 public constant UNIV3_POOL_INIT_CODE_HASH = 0xe34f199b19b2b4f47f68442619d555527d244f78a3297ea89325f843f87b8b54;
     address public constant UNIV3_FACTORY = 0x1F98431c8aD98523631AE4a59f267346ea31F984;
-    uint24[4] univ3_fees = [uint24(100), 500, 3000, 10000];
+    uint24[4] univ3_fees = [uint24(100), 500, 3000, 10000];
```
[the remark this answers is not in the document]
Comment (GalloDaSballo) [reply contracts/OnChainPricingMainnet.sol:65]: 100%

Can prob do a pure function that returns the list of fees. Saves 2.1k + 100 * 3 = 2.4k

[the remark this answers is not in the document]
Comment (GalloDaSballo) [reply contracts/OnChainPricingMainnet.sol]: Update: See: https://github.com/GalloDaSballo/fair-selling/pull/24/commits/6ee21302fa2db9a1521fdfde24c698cc02d91dd8

**Comment by GalloDaSballo** [comment]
Before Release:
- Understand why minor difference between V2 and V3 Pricers

- Additional polish TBD

**Comment by GalloDaSballo** [comment]
For V4 / V5:

Pre-sort given liquidity
Use Liquidity to calculate Certainty Ratio 
Use Liquidity and Price to estimate attack cost (cost of moving all valid Liquidity Sources to X price)
Use Certainty + Attack Cost to calculate "safe swap" amount of want that is economically unviable to front-run
Use Certainty + Attack Cost to calculate min-number of TWAP blocks before a bigger trade could be done (self-defending swapper)

**Comment by GalloDaSballo** [comment]
## Gas Profiles // MAY CHANGE
Screenshot 2022-07-27 at 04 30 34

Screenshot 2022-07-27 at 04 30 43


Screenshot 2022-07-27 at 04 30 49


## Coverage // WILL CHANGE
Screenshot 2022-07-27 at 04 31 05



Screenshot 2022-07-27 at 04 31 11

Screenshot 2022-07-27 at 04 31 21

Screenshot 2022-07-27 at 04 31 27



Screenshot 2022-07-27 at 04 31 34

Screenshot 2022-07-27 at 04 31 55


Tbh we only care about our own contracts coverage, but increasing coverage to guarantee basic functionality on libraries should be beneficial for us as well

**Comment by GalloDaSballo** [comment]
Also no worries on coverage for V2, let's just raise V3 and keep V2 as baseline behaviour

**Comment by rayeaster** [comment]
```
================================================================================================================= Coverage =================================================================================================================

  contract: BalancerSwapSimulator - 58.7%
    BalancerMath.div - 100.0%
    BalancerFixedPoint.powUp - 79.2%
    BalancerFixedPoint.add - 75.0%
    BalancerFixedPoint.mulDown - 75.0%
    BalancerFixedPoint.sub - 75.0%
    BalancerMath.divDown - 75.0%
    BalancerMath.mul - 75.0%
    BalancerSwapSimulator.calcOutGivenIn - 75.0%
    BalancerStableMath._getTokenBalanceGivenInvariantAndAllOtherBalances - 65.0%
    BalancerFixedPoint.divDown - 62.5%
    BalancerFixedPoint.divUp - 62.5%
    BalancerStableMath._calculateInvariant - 61.4%
    BalancerFixedPoint.mulUp - 58.3%
    BalancerMath.divUp - 58.3%
    BalancerLogExpMath.exp - 57.4%
    BalancerLogExpMath.pow - 55.2%
    BalancerLogExpMath._ln - 48.7%

  contract: OnChainPricingMainnet - 87.2%
    OnChainPricingMainnet._checkPoolLiquidityAndBalances - 100.0%
    OnChainPricingMainnet._findTokenInBalancePool - 100.0%
    OnChainPricingMainnet._ifUniV3Token0Price - 100.0%
    OnChainPricingMainnet._simLoopAllUniV3Pools - 100.0%
    OnChainPricingMainnet.checkUniV3InRangeLiquidity - 100.0%
    OnChainPricingMainnet.getBalancerPriceAnalytically - 100.0%
    OnChainPricingMainnet.getUniPrice - 100.0%
    OnChainPricingMainnet.getUniV3PriceWithConnector - 100.0%
    OnChainPricingMainnet.pairForUniV2 - 100.0%
    OnChainPricingMainnet.sortUniV3Pools - 100.0%
    OnChainPricingMainnet.isPairSupported - 95.0%
    OnChainPricingMainnet.getBalancerQuoteWithinPoolAnalytcially - 91.7%
    OnChainPricingMainnet.getBalancerV2Pool - 88.4%
    OnChainPricingMainnet._useSinglePoolInUniV3 - 87.8%
    OnChainPricingMainnet._findOptimalSwap - 85.5%
    OnChainPricingMainnet.getBalancerPriceWithConnectorAnalytically - 66.7%
    OnChainPricingMainnet.setBalancerV2Simulator - 0.0%
    OnChainPricingMainnet.setUniV3Simulator - 0.0%

  contract: OnChainPricingMainnetLenient - 44.1%
    OnChainPricingMainnet._checkPoolLiquidityAndBalances - 100.0%
    OnChainPricingMainnet._ifUniV3Token0Price - 100.0%
    OnChainPricingMainnet._simLoopAllUniV3Pools - 100.0%
    OnChainPricingMainnet.checkUniV3InRangeLiquidity - 100.0%
    OnChainPricingMainnet.getBalancerPriceAnalytically - 100.0%
    OnChainPricingMainnet.pairForUniV2 - 100.0%
    OnChainPricingMainnet.sortUniV3Pools - 100.0%
    OnChainPricingMainnet._findTokenInBalancePool - 83.3%
    OnChainPricingMainnet.getUniPrice - 75.0%
    OnChainPricingMainnetLenient.setSlippage - 75.0%
    OnChainPricingMainnet._findOptimalSwap - 70.9%
    OnChainPricingMainnet.getBalancerQuoteWithinPoolAnalytcially - 66.7%
    OnChainPricingMainnet._useSinglePoolInUniV3 - 37.8%
    OnChainPricingMainnet.getBalancerV2Pool - 19.6%
    OnChainPricingMainnet.getBalancerPriceWithConnectorAnalytically - 0.0%
    OnChainPricingMainnet.getUniV3PriceWithConnector - 0.0%
    OnChainPricingMainnet.isPairSupported - 0.0%
    OnChainPricingMainnet.setBalancerV2Simulator - 0.0%
    OnChainPricingMainnet.setUniV3Simulator - 0.0%

  contract: OnChainSwapMainnet - 63.3%
    OnChainSwapMainnet.execSwapBalancerV2Batch - 75.0%
    OnChainSwapMainnet.execSwapBalancerV2Single - 75.0%
    OnChainSwapMainnet.execSwapCurve - 75.0%
    OnChainSwapMainnet.execSwapUniV2 - 75.0%
    OnChainSwapMainnet.execSwapUniV3 - 75.0%
    OnChainSwapMainnet.doOptimalSwapWithQuote - 74.7%
    Address.verifyCallResult - 37.5%
    OnChainSwapMainnet.setPricer - 0.0%
    OnChainSwapMainnet.setSwapSlippageTolerance - 0.0%

  contract: UniV3SwapSimulator - 79.1%
    SwapMath._getExactInNextPrice - 100.0%
    TickBitmap.nextInitializedTickWithinOneWord - 100.0%
    UniV3SwapSimulator._getAmountOutputDelta - 100.0%
    UniV3SwapSimulator._getLimitPrice - 100.0%
    UniV3SwapSimulator._getTargetPriceForSwapStep - 100.0%
    BitMath.mostSignificantBit - 97.2%
    TickMath.getSqrtRatioAtTick - 94.5%
    BitMath.leastSignificantBit - 91.5%
    UniV3SwapSimulator._stepInTick - 90.0%
    TickMath.getTickAtSqrtRatio - 85.0%
    FullMath.mulDiv - 83.3%
    LiquidityMath.addDelta - 83.3%
    SqrtPriceMath.getNextSqrtPriceFromInput - 83.3%
    FullMath.mulDivRoundingUp - 75.0%
    LowGasSafeMath.add - 75.0%
    SafeCast.toInt256 - 75.0%
    SafeCast.toUint160 - 75.0%
    UniV3SwapSimulator.checkInRangeLiquidity - 75.0%
    SqrtPriceMath.getAmount0Delta - 66.7%
    SqrtPriceMath.getAmount1Delta - 62.5%
    SwapMath.computeSwapStep - 58.6%
    UniV3SwapSimulator._getNextInitializedTick - 50.0%
    SqrtPriceMath.getNextSqrtPriceFromAmount1RoundingDown - 29.2%
    SqrtPriceMath.getNextSqrtPriceFromAmount0RoundingUp - 25.0%
    SqrtPriceMath.getNextSqrtPriceFromOutput - 0.0%
```

**Comment by GalloDaSballo** [comment]
Final Touches:
[ ] Understand why test crashes
[ ] BREAKING CHANGE - Change functions to view and refactor tests for that

**Pull request merged**